### PR TITLE
Added support for using IServiceProvider when configuring Rabbitmq

### PR DESCRIPTION
**Pull request by isdaniel**
**What this PR does / why we need it**:

This pull request is supported for configuring the connection and option using an `IServiceProvider` similar to how the `SqlServerHealthCheck` supports this for a connection string & Uri. 

So instead of this:

```
var provider = services.BuildServiceProvider();
var settings = services.GetRequiredService<IOptions>();
```

We need to set string type for Rabbitmq connection from heartbeat before.

```
services.AddHealthChecks().AddRabbitMQ("Connectionstring");
```

You can now do something like the code below.

```
//for connection string
services.AddHealthChecks().AddRabbitMQ(sp => services.GetRequiredService<IOptions>().Value.RabbitmqConnectionString);

//for uri string
services.AddHealthChecks().AddRabbitMQ(sp => services.GetRequiredService<IOptions>().Value.RabbitmqConnectionString);
```

**Special notes for your reviewer**:

This is based on the connectionStringFactory implementation in the SqlServerHealthCheck.

I had added some unit tests & functional tests for this feature, I passed by this picture below

image

**Does this PR introduce a user-facing change?**:

Please make sure you've completed the relevant tasks for this PR, out of the following list:

It extends but does not break the existing options for configuring a RabbitMQ HealthCheck

- [x] Code compiles correctly
- [x] Created/updated tests
- [x] Unit tests passing
- [x] End-to-end tests passing
- [ ] Extended the documentation
- [ ] Provided sample for the feature

### CODE REVIEW

**Comment by sungam3r** [comment]
@isdaniel I merged #912 and #913 that affect your PR. Please rebase it onto `master`.

**Comment by isdaniel** [comment]
@sungam3r I had edited my PR & tried to rebase onto master, I seldom do rebase I am not sure that is correct

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 1]
```diff
@@ -36,15 +36,7 @@ public static class RabbitMQHealthCheckBuilderExtensions
...
-            builder.Services
-                .AddSingleton(sp => new RabbitMQHealthCheck(new Uri(rabbitConnectionString), sslOption));
-
-            return builder.Add(new HealthCheckRegistration(
-                name ?? NAME,
-                sp => sp.GetRequiredService<RabbitMQHealthCheck>(),
-                failureStatus,
-                tags,
-                timeout));
+            return builder.AddRabbitMQ(new Uri(rabbitConnectionString),sslOption,name,failureStatus,tags,timeout);
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:39]:
```suggestion
            return builder.AddRabbitMQ(new Uri(rabbitConnectionString), sslOption, name, failureStatus, tags, timeout);
```

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 2]
```diff
@@ -191,5 +183,63 @@ public static class RabbitMQHealthCheckBuilderExtensions
                 tags,
                 timeout));
         }
+
+        /// <summary>
+        /// Add a health check for RabbitMQ services using connection string (amqp uri).
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:188]:
```suggestion
        /// Add a health check for RabbitMQ services using connection string (amqp uri) factory.
```

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 3]
```diff
@@ -182,7 +174,40 @@ public static class RabbitMQHealthCheckBuilderExtensions
...
+                failureStatus,
+                tags,
+                timeout));
+        }
+
+        /// <summary>
+        /// Add a health check for RabbitMQ services using connection string (amqp uri) factory.
+        /// </summary>
+        /// <param name="builder">The <see cref="IHealthChecksBuilder"/>.</param>
+        /// <param name="connectionFactoryFactory"> A factory function to provide the rabbitMQ connection Uri</param>
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:191]:
```suggestion
        /// <param name="connectionFactoryFactory">A factory function to provide the RabbitMQ connection string (amqp uri).</param>
```

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 4]
```diff
@@ -182,7 +174,40 @@ public static class RabbitMQHealthCheckBuilderExtensions
...
+        /// </summary>
+        /// <param name="builder">The <see cref="IHealthChecksBuilder"/>.</param>
+        /// <param name="connectionFactoryFactory">A factory function to provide the RabbitMQ connection string (amqp uri).</param>
+        /// <param name="name">The health check name. Optional. If <c>null</c> the type name 'rabbitmq' will be used for the name.</param>
+        /// <param name="failureStatus">
+        /// The <see cref="HealthStatus"/> that should be reported when the health check fails. Optional. If <c>null</c> then
+        /// the default status of <see cref="HealthStatus.Unhealthy"/> will be reported.
+        /// </param>
+        /// <param name="tags">A list of tags that can be used to filter sets of health checks. Optional.</param>
+        /// <param name="timeout">An optional System.TimeSpan representing the timeout of the check.</param>
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:198]:
```suggestion
        /// <param name="timeout">An optional <see cref="TimeSpan"/> representing the timeout of the check.</param>
```

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 5]
```diff
@@ -191,5 +183,63 @@ public static class RabbitMQHealthCheckBuilderExtensions
...
+        /// <param name="failureStatus">
+        /// The <see cref="HealthStatus"/> that should be reported when the health check fails. Optional. If <c>null</c> then
+        /// the default status of <see cref="HealthStatus.Unhealthy"/> will be reported.
+        /// </param>
+        /// <param name="tags">A list of tags that can be used to filter sets of health checks. Optional.</param>
+        /// <param name="timeout">An optional <see cref="TimeSpan"/> representing the timeout of the check.</param>
+        /// <returns>The <see cref="IHealthChecksBuilder"/>.</returns>
+        public static IHealthChecksBuilder AddRabbitMQ(
+            this IHealthChecksBuilder builder,
+            Func<IServiceProvider,Uri> connectionStringFactory,
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:202]:
```suggestion
            Func<IServiceProvider, Uri> connectionStringFactory,
```

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 6]
```diff
@@ -191,5 +183,63 @@ public static class RabbitMQHealthCheckBuilderExtensions
...
+        /// <param name="builder">The <see cref="IHealthChecksBuilder"/>.</param>
+        /// <param name="connectionFactoryFactory">A factory function to provide the RabbitMQ connection string (amqp uri).</param>
+        /// <param name="name">The health check name. Optional. If <c>null</c> the type name 'rabbitmq' will be used for the name.</param>
+        /// <param name="failureStatus">
+        /// The <see cref="HealthStatus"/> that should be reported when the health check fails. Optional. If <c>null</c> then
+        /// the default status of <see cref="HealthStatus.Unhealthy"/> will be reported.
+        /// </param>
+        /// <param name="tags">A list of tags that can be used to filter sets of health checks. Optional.</param>
+        /// <param name="timeout">An optional <see cref="TimeSpan"/> representing the timeout of the check.</param>
+        /// <returns>The <see cref="IHealthChecksBuilder"/>.</returns>
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:199]:
```suggestion
        /// <returns>The specified <paramref name="builder"/>.</returns>
```

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 7]
```diff
@@ -191,5 +216,30 @@ public static class RabbitMQHealthCheckBuilderExtensions
                 tags,
                 timeout));
         }
+
+        /// <summary>
+        /// Add a health check for RabbitMQ services using connection string (amqp uri).
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:221]:
```suggestion
        /// Add a health check for RabbitMQ services using connection string (amqp uri) factory.
```

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 8]
```diff
@@ -191,5 +216,30 @@ public static class RabbitMQHealthCheckBuilderExtensions
...
+        /// </summary>
+        /// <param name="builder">The <see cref="IHealthChecksBuilder"/>.</param>
+        /// <param name="connectionFactoryFactory"> A factory function to provide the rabbitMQ connection string</param>
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:224]:
```suggestion
        /// <param name="connectionFactoryFactory">A factory function to provide the RabbitMQ connection string (amqp uri).</param>
```

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 9]
```diff
@@ -191,5 +216,30 @@ public static class RabbitMQHealthCheckBuilderExtensions
...
+        /// <param name="name">The health check name. Optional. If <c>null</c> the type name 'rabbitmq' will be used for the name.</param>
+        /// <param name="failureStatus">
+        /// The <see cref="HealthStatus"/> that should be reported when the health check fails. Optional. If <c>null</c> then
+        /// the default status of <see cref="HealthStatus.Unhealthy"/> will be reported.
+        /// </param>
+        /// <param name="tags">A list of tags that can be used to filter sets of health checks. Optional.</param>
+        /// <param name="timeout">An optional System.TimeSpan representing the timeout of the check.</param>
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:231]:
```suggestion
        /// <param name="timeout">An optional <see cref="TimeSpan"/> representing the timeout of the check.</param>
```

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 10]
```diff
@@ -191,5 +216,30 @@ public static class RabbitMQHealthCheckBuilderExtensions
...
+        /// <returns>The <see cref="IHealthChecksBuilder"/>.</returns>
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:232]:
```suggestion
        /// <returns>The specified <paramref name="builder"/>.</returns>
```

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 11]
```diff
@@ -191,5 +216,30 @@ public static class RabbitMQHealthCheckBuilderExtensions
...
+        public static IHealthChecksBuilder AddRabbitMQ(
+            this IHealthChecksBuilder builder,
+            Func<IServiceProvider,string> connectionStringFactory,
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:235]:
```suggestion
            Func<IServiceProvider, string> connectionStringFactory,
```

File: src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs [hunk 12]
```diff
@@ -191,5 +216,30 @@ public static class RabbitMQHealthCheckBuilderExtensions
...
+            SslOption? sslOption = default,
+            string? name = default,
+            HealthStatus? failureStatus = default,
+            IEnumerable<string>? tags = default,
+            TimeSpan? timeout = default)
+        {
+            return  builder.AddRabbitMQ((sp)=> new Uri(connectionStringFactory(sp)),sslOption,name,failureStatus,tags,timeout);
```
Comment (sungam3r) [inline src/HealthChecks.Rabbitmq/DependencyInjection/RabbitMQHealthCheckBuilderExtensions.cs:242]:
```suggestion
            return  builder.AddRabbitMQ(sp => new Uri(connectionStringFactory(sp)), sslOption, name, failureStatus, tags, timeout);
```

File: test/HealthChecks.RabbitMQ.Tests/DependencyInjection/RegistrationTests.cs [hunk 13]
```diff
@@ -58,5 +58,53 @@ public void add_named_health_check_when_properly_configured()
             result.Status.Should().Be(HealthStatus.Unhealthy);
             result.Exception.GetType().Should().Be(typeof(ObjectDisposedException));
         }
+
+        [Fact]
+        public void add_named_health_check_with_connection_string_factory_by_iServiceProvider_registered()
+        {
+            var services = new ServiceCollection();
+            var customCheckName = "my-" + DEFAULT_CHECK_NAME;
+            services.AddSingleton(new RabbitMqSetting(){
```
Comment (sungam3r) [inline test/HealthChecks.RabbitMQ.Tests/DependencyInjection/RegistrationTests.cs:67]:
```suggestion
            services.AddSingleton(new RabbitMqSetting
            {
```

File: test/HealthChecks.RabbitMQ.Tests/DependencyInjection/RegistrationTests.cs [hunk 14]
```diff
@@ -58,5 +58,53 @@ public void add_named_health_check_when_properly_configured()
...
+        public void add_named_health_check_with_connection_string_factory_by_iServiceProvider_registered()
+        {
+            var services = new ServiceCollection();
+            var customCheckName = "my-" + DEFAULT_CHECK_NAME;
+            services.AddSingleton(new RabbitMqSetting(){
+                ConnectionString = FAKE_CONNECTION_STRING
+            });
+
+            services.AddHealthChecks()
+                .AddRabbitMQ((sp)=> sp.GetRequiredService<RabbitMqSetting>().ConnectionString, name: customCheckName);
```
Comment (sungam3r) [inline test/HealthChecks.RabbitMQ.Tests/DependencyInjection/RegistrationTests.cs:72]:
```suggestion
                .AddRabbitMQ(sp => sp.GetRequiredService<RabbitMqSetting>().ConnectionString, name: customCheckName);

```

File: test/HealthChecks.RabbitMQ.Tests/DependencyInjection/RegistrationTests.cs [hunk 15]
```diff
@@ -58,5 +58,53 @@ public void add_named_health_check_when_properly_configured()
...
+
+            var serviceProvider = services.BuildServiceProvider();
```
Comment (sungam3r) [inline test/HealthChecks.RabbitMQ.Tests/DependencyInjection/RegistrationTests.cs:74]:
```suggestion
            using var serviceProvider = services.BuildServiceProvider();

```

File: test/HealthChecks.RabbitMQ.Tests/DependencyInjection/RegistrationTests.cs [hunk 16]
```diff
@@ -58,5 +58,53 @@ public void add_named_health_check_when_properly_configured()
...
+            registration.Name.Should().Be(customCheckName);
+            check.GetType().Should().Be(typeof(RabbitMQHealthCheck));
+        }
+
+        [Fact]
+        public void add_named_health_check_with_uri_string_factory_by_iServiceProvider_registered()
+        {
+            var services = new ServiceCollection();
+            var customCheckName = "my-" + DEFAULT_CHECK_NAME;
+            services.AddSingleton(new RabbitMqSetting(){
```
Comment (sungam3r) [inline test/HealthChecks.RabbitMQ.Tests/DependencyInjection/RegistrationTests.cs:89]:
```suggestion
            services.AddSingleton(new RabbitMqSetting
            {
```

File: test/HealthChecks.RabbitMQ.Tests/DependencyInjection/RegistrationTests.cs [hunk 17]
```diff
@@ -58,5 +58,53 @@ public void add_named_health_check_when_properly_configured()
...
+                ConnectionString = FAKE_CONNECTION_STRING
+            });
+
+            services.AddHealthChecks()
+                .AddRabbitMQ((sp)=> new Uri(sp.GetRequiredService<RabbitMqSetting>().ConnectionString), name: customCheckName);
```
Comment (sungam3r) [inline test/HealthChecks.RabbitMQ.Tests/DependencyInjection/RegistrationTests.cs:94]:
```suggestion
                .AddRabbitMQ(sp => new Uri(sp.GetRequiredService<RabbitMqSetting>().ConnectionString), name: customCheckName);
```

File: test/HealthChecks.RabbitMQ.Tests/DependencyInjection/RegistrationTests.cs [hunk 18]
```diff
@@ -58,5 +58,53 @@ public void add_named_health_check_when_properly_configured()
...
+
+            var serviceProvider = services.BuildServiceProvider();
```
Comment (sungam3r) [inline test/HealthChecks.RabbitMQ.Tests/DependencyInjection/RegistrationTests.cs:96]:
```suggestion
            using var serviceProvider = services.BuildServiceProvider();
```

File: test/HealthChecks.RabbitMQ.Tests/DependencyInjection/RegistrationTests.cs [hunk 19]
```diff
@@ -58,5 +58,53 @@ public void add_named_health_check_when_properly_configured()
...
+
+            var registration = options.Value.Registrations.First();
+            var check = registration.Factory(serviceProvider);
+
+            registration.Name.Should().Be(customCheckName);
+            check.GetType().Should().Be(typeof(RabbitMQHealthCheck));
+        }
+    }
+
+    public class RabbitMqSetting{
```
Comment (sungam3r) [inline test/HealthChecks.RabbitMQ.Tests/DependencyInjection/RegistrationTests.cs:107]:
```suggestion
    public class RabbitMqSetting
    {
```

File: test/HealthChecks.RabbitMQ.Tests/DependencyInjection/RegistrationTests.cs [hunk 20]
```diff
@@ -58,5 +58,57 @@ public void add_named_health_check_when_properly_configured()
...
+            var services = new ServiceCollection();
+            var customCheckName = "my-" + DEFAULT_CHECK_NAME;
+            services.AddSingleton(new RabbitMqSetting
+            {
+                ConnectionString = FAKE_CONNECTION_STRING
+            });
+
+            services.AddHealthChecks()
+                .AddRabbitMQ(sp => sp.GetRequiredService<RabbitMqSetting>().ConnectionString, name: customCheckName);
+
```
Comment (sungam3r) [inline test/HealthChecks.RabbitMQ.Tests/DependencyInjection/RegistrationTests.cs:74]:
```suggestion

```

File: test/HealthChecks.RabbitMQ.Tests/Functional/RabbitHealthCheckTests.cs [hunk 21]
```diff
@@ -241,5 +241,36 @@ public async Task be_not_crash_on_startup_when_rabbitmq_is_down_at_startup()
...
+        {
+            var connectionString = @"amqp://localhost:5672";
+
+            var webHostBuilder = new WebHostBuilder()
+            .UseStartup<DefaultStartup>()
+            .ConfigureServices(services =>
+            {
+                services
+                    .AddHealthChecks()
+                    .AddRabbitMQ(sp=> new Uri(connectionString),tags: new string[] { "rabbitmq" });
```
Comment (sungam3r) [inline test/HealthChecks.RabbitMQ.Tests/Functional/RabbitHealthCheckTests.cs:256]:
```suggestion
                    .AddRabbitMQ(sp => new Uri(connectionString), tags: new string[] { "rabbitmq" });

```

File: test/HealthChecks.RabbitMQ.Tests/Functional/RabbitHealthCheckTests.cs [hunk 22]
```diff
@@ -241,5 +241,36 @@ public async Task be_not_crash_on_startup_when_rabbitmq_is_down_at_startup()
...
+
+            var server = new TestServer(webHostBuilder);
+
+            var response = await server.CreateRequest($"/health")
+                .GetAsync();
+
+            response.StatusCode
+                .Should().Be(HttpStatusCode.OK);
+        }
+
```
Comment (sungam3r) [inline test/HealthChecks.RabbitMQ.Tests/Functional/RabbitHealthCheckTests.cs:274]:
```suggestion
```

**Pull request merged**